Build: move dependency transpilation config to calypso-build

jsnajdr opened the pull request:
Alternative to #33492 which still uses the special Babel config for transpiling dependencies, but moves it to the `calypso-build` package.

This PR, unlike #33492, doesn't attempt to transpile both app and `node_modules` code with the same Babel config. It keeps the separation and shouldn't cause any changes in output -- just moves code to a different location. That involves less risk in my opinion.

Also, this approach is in line with what Create React App does.

After merging this, `calypso-build` should start transpiling dependencies out of the box, without any configuration. Jetpack can just upgrade to the new version and be done.

**How to test:**
I'm not sure what's the right way to test if dependencies are still transpiled correctly. Load the calypso.live build in IE11?

jsnajdr committed:
Build: move dependency transpilation config to calypso-build

Alternative to #33492 which still uses the special Babel config for transpiling dependencies,
but moves it to the `calypso-build` package.

This PR, unlike #33492, doesn't attempt to transpile both app and `node_modules` code with the
same Babel config. It keeps the separation and shouldn't cause any changes in output -- just
moves code to a different location.

Also, this approach is in line with what Create React App does.

After merging this, `calypso-build` should start transpiling dependencies out of the box, without
any configuration. Jetpack can just upgrade to the new version and be done.

blowery commented on packages/calypso-build/webpack.config.js:
```diff
@@ -112,6 +112,12 @@ function getWebpackConfig(
 				} ),
+				TranspileConfig.loader( {
+					cacheDirectory: true,
+					include: shouldTranspileDependency,
```
We'll want to expose this to the user of calypso-build, as they may have dependencies that require transpilation that calypso does not.

jsnajdr replied:
Do users need to only expand the whitelist with new entries, or to also exclude some existing ones?

blowery commented on packages/calypso-build/webpack/util.js:
```diff
@@ -67,4 +67,50 @@ function IncrementalProgressPlugin() {
+	// general form is <package-name>/.
+	// The trailing slash makes sure we're not matching these as prefixes
+	// In some cases we do want prefix style matching (lodash. for lodash.assign)
+	'@github/webauthn-json/',
```
Would now be a good time to extend this to support semver?

jsnajdr replied:
The idea is to have semver specifiers like `debug@^3` in the whitelist and be able to read the `package.json` for a tested package and match the semver version?

That seems a big enough topic to warrant its own PR. We probably want to keep some cache of "file-path -> package-version" mappings, so that we don't do the look up for every file (every package can include a large number of contained files and webpack sees each sub-path separately).

blowery commented on packages/calypso-build/webpack/util.js:
```diff
@@ -67,4 +67,50 @@ function IncrementalProgressPlugin() {
+	'react-spring/',
+	'regenerate-unicode-properties/',
+	'regexpu-core/',
+	'striptags',
```
oop, this one should have a trailing `/`

jsnajdr replied:
If we go the semver route, and start exposing API for expanding the whitelist, then we should get rid of the trailing slashes and patch the matching algorithm instead.

jsnajdr commented:
I rebased this PR onto the latest master and it's still ready for review.

One open question is to whether to allow to customize the whitelist of transpiled dependencies and how.

We can also take the CRA route and start transpiling all `node_modules` dependencies except `@babel/runtime`, as they do.

@blowery @ockham @sgomes let me know if you have any concerns.

sgomes reviewed (commented):
No concerns from my end. This makes sense to live in `calypso-build`, and I don't see an immediate need to make the transpilation list configurable, as all projects making use of `calypso-build` presumably need (or can benefit from) the same configuration there.

sirreal commented:
> How to test:
> I'm not sure what's the right way to test if dependencies are still transpiled correctly. Load the calypso.live build in IE11?

I'd expect the es6 fallback build linter to catch issues.

sirreal reviewed (approved):
I haven't tested, but I looked it over and this makes sense. Thanks!

sirreal commented:
Will you add a changelog entry to calypso-build?

jsnajdr commented:
I rebased this PR to pickup the changes in CoreJS 3 upgrade (#38660) and added the changelog entry requested by @sirreal. I'll merge this after the CI checks come back green.

ockham commented:
Thanks all!
